### PR TITLE
Add contributing docs to aid documenting CA and IDE rules in future

### PR DESCRIPTION
**Pull request by mavasani**

### CODE REVIEW

**Comment by mavasani** [comment]
@gewarren @mikadumont for review.

File: Contribute/dotnet/dotnet-contribute.md
Comment (gewarren) [inline Contribute/dotnet/dotnet-contribute.md:86]: Probably better to switch steps 2 and 3 since they can't add entries to the TOC/indexes until the rule file has been created.

File: Contribute/dotnet/dotnet-contribute.md
Comment (gewarren) [inline Contribute/dotnet/dotnet-contribute.md:87]: Since these steps aren't an ordered procedure, they should be bulleted instead of numbered.

File: Contribute/dotnet/dotnet-contribute.md
Comment (gewarren) [inline Contribute/dotnet/dotnet-contribute.md:97]:
```suggestion
```

File: Contribute/dotnet/dotnet-contribute.md
Comment (gewarren) [inline Contribute/dotnet/dotnet-contribute.md:98]:
```suggestion
> Perform a repo search in the `dotnet/docs` repo for a known, existing CA rule ID to identify potential places that might need an update. For example, see <https://github.com/dotnet/docs/search?q=CA2000>.
```

File: Contribute/dotnet/dotnet-contribute.md
Comment (gewarren) [inline Contribute/dotnet/dotnet-contribute.md:83]:
```suggestion
Please follow the following steps to contribute documentation for code quality analysis rules to the [dotnet/docs](https://github.com/dotnet/docs) repo:
```

File: Contribute/dotnet/dotnet-contribute.md
Comment (gewarren) [inline Contribute/dotnet/dotnet-contribute.md:102]:
```suggestion
Please follow the following steps to contribute documentation for code style analysis rules to the [dotnet/docs](https://github.com/dotnet/docs) repo:
```

File: Contribute/dotnet/dotnet-contribute.md
Comment (gewarren) [inline Contribute/dotnet/dotnet-contribute.md:104]:
```suggestion
1. Determine `Rule ID` and code style options, if any: Ensure that you know the 'IDExxxx' rule ID and code style options for the rule to be documented. This means either the IDE analyzer has been merged into the [dotnet/roslyn](https://github.com/dotnet/roslyn) repo or you have an open PR with an approved ID and code style options that have been assigned to the rule.
```

File: Contribute/dotnet/dotnet-contribute.md
Comment (gewarren) [inline Contribute/dotnet/dotnet-contribute.md:105]:
```suggestion
2. Determine rule `subcategory`: IDE rules have category `Style`. Identify the rule subcategory by reading through the introductory section of [Code style rules](/dotnet/fundamentals/code-analysis/style-rules/index). Most of the IDE code style rules are under the `Language` subcategory.
```

File: Contribute/dotnet/dotnet-contribute.md
Comment (gewarren) [inline Contribute/dotnet/dotnet-contribute.md:108]:
```suggestion
  - Otherwise, determine if the rule belongs to an existing preference group or needs a new preference group.
```

File: Contribute/dotnet/dotnet-contribute.md
Comment (gewarren) [inline Contribute/dotnet/dotnet-contribute.md:109]: Switch steps 4 and 5, and bullet the subpoints of step 4 (which is moving to step 5).

File: Contribute/dotnet/dotnet-contribute.md
Comment (gewarren) [inline Contribute/dotnet/dotnet-contribute.md:110]:
```suggestion
  1. In [table of contents](https://github.com/dotnet/docs/blob/master/docs/fundamentals/toc.yml) file under the rule's sub-category and preferences list. For example, for an IDE code style rule with `Language` subcategory and `Modifier preferences` group, search for the term `- name: Language rules` and a child node `- name: Modifier preferences` in the file and add an entry to the list. If either the subcategory list or the preferences list under the subcategory does not exist, create a new one under `- name: Code style rules`.
```

File: Contribute/dotnet/dotnet-contribute.md
Comment (gewarren) [inline Contribute/dotnet/dotnet-contribute.md:111]:
```suggestion
  2. In the [code style option-based index](https://github.com/dotnet/docs/blob/master/docs/fundamentals/code-analysis/style-rules/language-rules.md) file.
```

File: Contribute/dotnet/dotnet-contribute.md
Comment (gewarren) [inline Contribute/dotnet/dotnet-contribute.md:112]:
```suggestion
  3. In the [rule ID-based index](https://github.com/dotnet/docs/blob/master/docs/fundamentals/code-analysis/style-rules/index.md) file.
```

File: Contribute/dotnet/dotnet-contribute.md
Comment (gewarren) [inline Contribute/dotnet/dotnet-contribute.md:121]:
```suggestion
```

File: Contribute/dotnet/dotnet-contribute.md
Comment (gewarren) [inline Contribute/dotnet/dotnet-contribute.md:122]:
```suggestion
> Perform a repo search in the `dotnet/docs` repo, for a known, existing IDE rule and/or code style option(s) to identify potential places that might need an update. For example, see <https://github.com/dotnet/docs/search?q=IDE0011> and <https://github.com/dotnet/docs/search?q=csharp_prefer_braces>.
```

**Review by gewarren — APPROVED** [review]
Thanks for adding this!

**Comment by mavasani** [comment]
Thanks @gewarren - I addressed your feedback. Who will be the right person to merge this PR?

**Comment by gewarren** [comment]
When you're ready to merge, you can just pound sign-off. But looking at the [preview](https://apidrop.visualstudio.com/Content%20CI/_build/results?buildId=166965&view=results), I think it might be good to promote the "Contribute analyzer docs" to top-level, or even move it into a separate file, since it's so specialized. @tdykstra what's your recommendation?

**Comment by tdykstra** [comment]
I would put it in a separate file and add it to the toc.  If it does stay in the current file it needs to be moved because it comes between the discussion of sample folder structure and the example illustrating the discussed guidelines.

File: Contribute/dotnet/dotnet-contribute-code-analysis.md
Comment (tdykstra) [inline Contribute/dotnet/dotnet-contribute-code-analysis.md:7]:
```suggestion
ms.date: 10/09/2020
```

**Comment by mavasani** [comment]
#sign-off

**Comment by gewarren** [comment]
Actually there is some indentation problems. https://review.docs.microsoft.com/en-us/contribute/dotnet/dotnet-contribute-code-analysis?branch=pr-en-us-443. I'll add suggestions. #hold-off.

File: Contribute/dotnet/dotnet-contribute-code-analysis.md
Comment (gewarren) [inline Contribute/dotnet/dotnet-contribute-code-analysis.md:26]:
```suggestion
   1. Clone an existing CA rule file under [root](https://github.com/dotnet/docs/blob/master/docs/fundamentals/code-analysis/quality-rules) folder, say `ca1000.md`, and rename it.
```

File: Contribute/dotnet/dotnet-contribute-code-analysis.md
Comment (gewarren) [inline Contribute/dotnet/dotnet-contribute-code-analysis.md:27]:
```suggestion
   2. Update the content of the file appropriately.
```

File: Contribute/dotnet/dotnet-contribute-code-analysis.md
Comment (gewarren) [inline Contribute/dotnet/dotnet-contribute-code-analysis.md:29]:
```suggestion
   - In [table of contents](https://github.com/dotnet/docs/blob/master/docs/fundamentals/toc.yml) file under the rule category list. For example, for a CA rule with `Performance` category, search for the term `- name: Performance rules` in the file and add an entry to the list. If none exists, create a new category list under `- name: Code quality rules`.
```

File: Contribute/dotnet/dotnet-contribute-code-analysis.md
Comment (gewarren) [inline Contribute/dotnet/dotnet-contribute-code-analysis.md:30]:
```suggestion
   - In [top-level rule index](https://github.com/dotnet/docs/blob/master/docs/fundamentals/code-analysis/quality-rules/index.md) file.
```

File: Contribute/dotnet/dotnet-contribute-code-analysis.md
Comment (gewarren) [inline Contribute/dotnet/dotnet-contribute-code-analysis.md:31]:
```suggestion
   - In `category` based rule index file:
```

File: Contribute/dotnet/dotnet-contribute-code-analysis.md
Comment (gewarren) [inline Contribute/dotnet/dotnet-contribute-code-analysis.md:32]:
```suggestion
     1. Identify the category based index file under [root](https://github.com/dotnet/docs/blob/master/docs/fundamentals/code-analysis/quality-rules) folder. For example, for a CA rule with `Performance` category, this file is named `performance-warnings.md`. If none exists, create a new category based index file by cloning an existing one.
```

File: Contribute/dotnet/dotnet-contribute-code-analysis.md
Comment (gewarren) [inline Contribute/dotnet/dotnet-contribute-code-analysis.md:33]:
```suggestion
     2. Add an entry for the rule ID in this file.
```

File: Contribute/dotnet/dotnet-contribute-code-analysis.md
Comment (gewarren) [inline Contribute/dotnet/dotnet-contribute-code-analysis.md:45]:
```suggestion
   - If the rule has known, associated code style option(s), the preferences group will match the `OptionGroup` specified in the code style option declaration.
```

File: Contribute/dotnet/dotnet-contribute-code-analysis.md
Comment (gewarren) [inline Contribute/dotnet/dotnet-contribute-code-analysis.md:46]:
```suggestion
   - Otherwise, determine if the rule belongs to an existing preference group or needs a new preference group.
```

File: Contribute/dotnet/dotnet-contribute-code-analysis.md
Comment (gewarren) [inline Contribute/dotnet/dotnet-contribute-code-analysis.md:48]:
```suggestion
   1. Clone an existing IDE rule file under [root](https://github.com/dotnet/docs/blob/master/docs/fundamentals/code-analysis/style-rules) folder, say `ide0011.md`, and rename it.
```

File: Contribute/dotnet/dotnet-contribute-code-analysis.md
Comment (gewarren) [inline Contribute/dotnet/dotnet-contribute-code-analysis.md:49]:
```suggestion
   2. Update the content of the file appropriately.
```

File: Contribute/dotnet/dotnet-contribute-code-analysis.md
Comment (gewarren) [inline Contribute/dotnet/dotnet-contribute-code-analysis.md:51]:
```suggestion
   - In [table of contents](https://github.com/dotnet/docs/blob/master/docs/fundamentals/toc.yml) file under the rule's sub-category and preferences list. For example, for an IDE code style rule with `Language` subcategory and `Modifier preferences` group, search for the term `- name: Language rules` and a child node `- name: Modifier preferences` in the file and add an entry to the list. If either the subcategory list or the preferences list under the subcategory does not exist, create a new one under `- name: Code style rules`.
```

File: Contribute/dotnet/dotnet-contribute-code-analysis.md
Comment (gewarren) [inline Contribute/dotnet/dotnet-contribute-code-analysis.md:52]:
```suggestion
   - In the [code style option-based index](https://github.com/dotnet/docs/blob/master/docs/fundamentals/code-analysis/style-rules/language-rules.md) file.
```

File: Contribute/dotnet/dotnet-contribute-code-analysis.md
Comment (gewarren) [inline Contribute/dotnet/dotnet-contribute-code-analysis.md:53]:
```suggestion
   - In the [rule ID-based index](https://github.com/dotnet/docs/blob/master/docs/fundamentals/code-analysis/style-rules/index.md) file.
```

File: Contribute/dotnet/dotnet-contribute-code-analysis.md
Comment (gewarren) [inline Contribute/dotnet/dotnet-contribute-code-analysis.md:54]:
```suggestion
   - In `preferences group` based rule index file:
```

File: Contribute/dotnet/dotnet-contribute-code-analysis.md
Comment (gewarren) [inline Contribute/dotnet/dotnet-contribute-code-analysis.md:55]:
```suggestion
     1. Identify the preferences group based index file under [root](https://github.com/dotnet/docs/blob/master/docs/fundamentals/code-analysis/style-rules) folder. For example, for an IDE rule for `Modifier preferences`, this file is named `modifier-preferences.md`. If none exists, create a new preferences group based index file by cloning an existing one.
```

File: Contribute/dotnet/dotnet-contribute-code-analysis.md
Comment (gewarren) [inline Contribute/dotnet/dotnet-contribute-code-analysis.md:56]:
```suggestion
     2. Add an entry for the rule ID in this file.
```

**Review by gewarren — COMMENTED** [review]
To fix the indentation. Paragraphs/bullets as part of a numbered item must be indented 3 spaces.

**Comment by mavasani** [comment]
Thanks @gewarren, the indenting looks good now! #sign-off

**Comment by gewarren** [comment]
@Jim-Parker Can you merge this when you get a chance?

**Comment by mavasani** [comment]
@Jim-Parker @gewarren When can we expect this to go live after being merged? I am planning to update analyzer suggestion issue templates in roslyn-analyzers and roslyn repo to add a link to this doc and also plan to send it out to our team to help the docs to always state up-to-date.

**Comment by Jim-Parker** [comment]
@mavasani please add the required metadata per [this metadata guidance](https://review.docs.microsoft.com/en-us/help/contribute/metadata-attributes?branch=master#required-metadata). 
Also, looking over the instructions I would suggest removing "Learn how to" from the H1, so the title of the article becomes "Contribute docs for .NET code analysis rules to the .NET docs repository". That seems more inline with the context
Thanks
#hold-off

**Comment by mavasani** [comment]
Done, thanks!

**Comment by mavasani** [comment]
#sign-off

**Pull request merged**
